File: ios/Stockbook/Features/Settings/SettingsScreen.swift
import SwiftUI

/// Settings: the shop's name, the two things it reads and bills in, and a way
/// through to the backup handoff.
///
/// It used to hold that handoff too, and ran to one long scroll where the two
/// controls an owner touches occasionally sat above two cards they use once a
/// year. Export and import now live behind a single row in `BackupScreen`, and
/// what is left fits without scrolling on most phones.
struct SettingsScreen: View {
    @Environment(StockbookStore.self) private var store
    @Environment(AppRouter.self) private var router

    @State private var ownerName = ""
    @State private var shopAddress = ""
    @State private var seeded = false
    @State private var confirmingPhotoRemoval = false

    private var settings: Settings { store.settings }
    private var products: [Product] { store.products }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(title: Loc.settings) {
                Button(Loc.done) { router.showingSettings = false }
                    .buttonStyle(GhostButtonStyle(fontSize: 12.5))
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    thisPhone
                    languageAndCurrency
                    theme
                    backupRow
                    photoStorageRow
                    #if DEBUG
                    startAgain
                    #endif
                }
                .padding(.horizontal, Metrics.screenPadding)
                .padding(.bottom, 18)
            }
        }
        .background(Nocturne.bg.ignoresSafeArea())
        .ignoresSafeArea(.keyboard, edges: .bottom)
        .onAppear(perform: seed)
    }

    // MARK: This phone

    private var thisPhone: some View {
        VStack(alignment: .leading, spacing: 0) {
            Kicker(Loc.thisPhone).padding(.bottom, 8)

            VStack(alignment: .leading, spacing: 10) {
                NocturneField(
                    label: Loc.businessOwner,
                    placeholder: Loc.businessOwnerName,
                    text: $ownerName
                )
                .onChange(of: ownerName) { _, new in
                    store.setOwnerName(new)
                }

                // Free text with line breaks in it, and taller than one line
                // because that is how an address is written. Nothing parses it —
                // the statement copies it out exactly as typed.
                VStack(alignment: .leading, spacing: 5) {
                    Text(Loc.shopAddress).nocturneText(.fieldLabel)
                    TextEditor(text: $shopAddress)
                        .font(NocturneType.inter(14))
                        .foregroundStyle(Nocturne.text)
                        .scrollContentBackground(.hidden)
                        .tint(Nocturne.accent)
                        .frame(height: 84)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 4)
                        .background(Nocturne.bg, in: RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                        .hairline(radius: Metrics.controlRadius)
                        .overlay(alignment: .topLeading) {
                            if shopAddress.isEmpty {
                                Text(Loc.shopAddressHint)
                                    .font(NocturneType.inter(14))
                                    .foregroundStyle(Nocturne.neutral500)
                                    .padding(.horizontal, 11)
                                    .padding(.vertical, 12)
                                    .allowsHitTesting(false)
                            }
                        }
                        .onChange(of: shopAddress) { _, new in
                            store.setShopAddress(new)
                        }
                    Text(Loc.shopAddressNote).nocturneText(.meta)
                }

                HStack(spacing: 10) {
                    stat(Loc.productsStat, products.count)
                    stat(Loc.billsStat, store.bills.count)
                    stat(Loc.customersStat, store.customers().count)
                }
            }
            .padding(12)
            .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
            .padding(.bottom, 20)
        }
    }

    private func stat(_ label: String, _ value: Int) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(label)
                .font(NocturneType.inter(11))
                .foregroundStyle(Nocturne.neutral500)
            Text(String(value)).font(NocturneType.inter(17))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    // MARK: What it reads and bills in

    /// One card, two dropdowns. They are the same kind of decision — pick one
    /// from a short list, applies everywhere at once — so they read better as a
    /// pair than as two sections with a paragraph each.
    private var languageAndCurrency: some View {
        VStack(alignment: .leading, spacing: 0) {
            Kicker(Loc.languageAndCurrency).padding(.bottom, 8)

            VStack(alignment: .leading, spacing: 10) {
                LanguageField(
                    label: Loc.languageSection,
                    language: Binding(
                        get: { settings.language },
                        set: { store.setLanguage($0) }
                    )
                )

                CurrencyField(
                    label: Loc.currencySection,
                    currency: Binding(
                        get: { settings.currency },
                        set: { store.setCurrency($0) }
                    )
                )
            }
            .padding(12)
            .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
            .padding(.bottom, 8)

            // The one caveat that cannot be discovered by trying it: the numbers
            // do not move when the symbol does.
            Text(Loc.currencyNote)
                .font(NocturneType.inter(12))
                .foregroundStyle(Nocturne.neutral500)
                .lineSpacing(3)
                .padding(.bottom, 20)
        }
    }

    // MARK: How it looks

    /// Two pills rather than a dropdown. There are exactly two answers, both
    /// worth showing at once, and the choice is the sort somebody makes by
    /// looking at the result — a menu that hides the alternative behind a tap is
    /// the wrong shape for that.
    ///
    /// There is no "System". Following the phone would hand the decision to
    /// whoever set the phone up, who is not always the person behind the counter.
    private var theme: some View {
        VStack(alignment: .leading, spacing: 0) {
            Kicker(Loc.themeSection).padding(.bottom, 8)

            HStack(spacing: 6) {
                ForEach(AppTheme.allCases) { option in
                    ChoicePill(
                        title: option.name(Loc),
                        icon: option == .light ? Icon.themeLight : Icon.themeDark,
                        selected: settings.theme == option
                    ) {
                        store.setTheme(option)
                    }
                }
            }
            .padding(.bottom, 20)
        }
    }

    // MARK: Move to another phone

    /// A row, not a section. The subtitle carries the backup state, because the
    /// standing reminder that nothing is backed up has to survive being folded
    /// away behind a tap.
    private var backupRow: some View {
        Button {
            router.showingBackup = true
        } label: {
            HStack(spacing: 11) {
                Glyph(settings.hasBackup ? Icon.backupDone : Icon.backupMissing, size: 20)
                    .foregroundStyle(settings.hasBackup ? Nocturne.accent : Nocturne.accent400)

                VStack(alignment: .leading, spacing: 2) {
                    Text(Loc.moveToAnotherPhone).nocturneText(.rowPrimary)
                    Text(backupState).nocturneText(.meta)
                }
                .frame(maxWidth: .infinity, alignment: .leading)

                Glyph(Icon.openRow, size: 12)
                    .foregroundStyle(Nocturne.neutral600)
            }
            .padding(13)
            .frame(maxWidth: .infinity)
            .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .padding(.bottom, 20)
    }

    /// What the photographs are costing, and what is missing.
    ///
    /// Storage that grows where nobody can see it is what gets an app deleted, so
    /// the figure is shown plainly. The missing count matters more than it looks:
    /// it is the only place a book that arrived from another phone says so, and
    /// without it an incomplete transfer is something the owner discovers a year
    /// later, one bill at a time.
    ///
    /// Absent entirely on a shop that has never taken one — an empty row about a
    /// feature somebody is not using is noise.
    @ViewBuilder
    private var photoStorageRow: some View {
        let named = store.photoIDsInUse()
        if !named.isEmpty {
            let photos = PhotoStore()
            let usage = photos.usage()
            let missing = named.filter { !photos.has(id: $0) }.count

            VStack(alignment: .leading, spacing: 10) {
                HStack(spacing: 11) {
                    Glyph(Icon.items, size: 20).foregroundStyle(Nocturne.accent400)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(Loc.photoStorage).nocturneText(.rowPrimary)
                        Text(
                            [
                                Loc.photosOnThisPhone(count: usage.count, size: Bytes.text(usage.bytes)),
                                missing > 0 ? Loc.photosMissing(missing) : nil
                            ].compactMap { $0 }.joined(separator: " · ")
                        )
                        .nocturneText(.meta)
                        .foregroundStyle(missing > 0 ? Nocturne.accent400 : Nocturne.neutral500)
                        .lineLimit(1)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                }

                if usage.count > 0 {
                    VStack(alignment: .leading, spacing: 6) {
                        Button(confirmingPhotoRemoval ? Loc.tapAgainToRemove : Loc.removeAllPhotos) {
                            if confirmingPhotoRemoval {
                                // The book keeps its ids. Only the files go —
                                // which is the same state a phone is in when a
                                // backup arrives ahead of its pictures, and the
                                // same thing brings them back.
                                photos.sweep(keeping: [])
                                confirmingPhotoRemoval = false
                            } else {
                                withAnimation(Metrics.quick) { confirmingPhotoRemoval = true }
                            }
                        }
                        .buttonStyle(GhostButtonStyle(
                            fontSize: 12,
                            tint: confirmingPhotoRemoval ? Nocturne.accent400 : Nocturne.neutral500,
                            horizontalPadding: 0
                        ))

                        Text(Loc.removeAllPhotosNote).nocturneText(.meta)
                    }
                }
            }
            .padding(13)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.cardRadius, style: .continuous))
            .padding(.bottom, 20)
        }
    }

    private var backupState: String {
        guard let exportedAt = settings.lastExportAt else { return Loc.notBackedUpYet }
        return Loc.backedUpOn(Loc.longDate(exportedAt))
    }

    // MARK: Start again

    /// Debug builds only.
    ///
    /// One tap, no confirmation step, and it clears every product, price and
    /// bill on the phone. That is the right shape for a developer resetting to
    /// first-run for the tenth time, and the wrong thing to leave sitting under
    /// Settings on a counter where the only copy of the shop lives. Release
    /// builds get here by deleting the app, which at least asks.
    ///
    /// `StockbookStore.startOver` itself is not conditional — the rule stays
    /// tested in both configurations, it just has no button.
    #if DEBUG
    private var startAgain: some View {
        VStack(alignment: .leading, spacing: 0) {
            Kicker(Loc.startAgain).padding(.bottom, 8)

            Button(Loc.startOver) {
                store.startOver()
                router.closeOverlays()
                router.showingSettings = false
                router.tab = .today
            }
            .buttonStyle(SecondaryButtonStyle(fullWidth: true, height: 42, fontSize: 13.5))
        }
    }
    #endif

    // MARK: Actions

    private func seed() {
        guard !seeded else { return }
        seeded = true
        ownerName = settings.ownerName
        shopAddress = settings.shopAddress
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/feature/settings/SettingsScreen.kt
package com.stockbook.app.feature.settings

import androidx.compose.foundation.clickable
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.BuildConfig
import com.stockbook.app.design.DropdownField
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Kicker
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneField
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.ScreenHeader
import com.stockbook.app.design.SecondaryButton
import com.stockbook.app.design.card
import com.stockbook.core.model.Currency
import com.stockbook.core.model.ShopState
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.Strings

/**
 * Settings: the shop's name, the two things it reads and bills in, and a way
 * through to the backup handoff.
 *
 * Short on purpose. The two controls an owner touches occasionally sit above a
 * single row for the two they use once a year.
 */
@Composable
fun SettingsScreen(
    state: ShopState,
    store: StockbookStore,
    strings: Strings,
    onClose: () -> Unit,
    onOpenBackup: () -> Unit,
    onStartOver: () -> Unit
) {
    var ownerName by remember(state.settings.ownerName) { mutableStateOf(state.settings.ownerName) }

    // Drawn as a sibling of the tab content inside the shell's Box, so without a
    // ground of its own the screen behind shows straight through it — and without
    // `statusBarsPadding` its header sits under the status bar.
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Nocturne.bg)
            .statusBarsPadding()
            .imePadding()
    ) {
        ScreenHeader(title = strings.settings) {
            GhostButton(strings.done, onClick = onClose)
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Metrics.screenPadding)
                .padding(bottom = 18.dp)
        ) {
            Kicker(strings.thisPhone, modifier = Modifier.padding(bottom = 8.dp))
            Column(modifier = Modifier.fillMaxWidth().card().padding(12.dp)) {
                NocturneField(
                    value = ownerName,
                    onValueChange = {
                        ownerName = it
                        store.setOwnerName(it)
                    },
                    label = strings.businessOwner,
                    placeholder = strings.businessOwnerName
                )
                Spacer(Modifier.height(10.dp))
                Row(modifier = Modifier.fillMaxWidth()) {
                    Stat(strings.productsStat, state.products.size, Modifier.weight(1f))
                    Stat(strings.billsStat, state.bills.count { !it.voided }, Modifier.weight(1f))
                    Stat(strings.customersStat, store.customers().size, Modifier.weight(1f))
                }
            }

            Spacer(Modifier.height(20.dp))
            Kicker(strings.languageAndCurrency, modifier = Modifier.padding(bottom = 8.dp))
            Column(modifier = Modifier.fillMaxWidth().card().padding(12.dp)) {
                DropdownField(
                    options = AppLanguage.entries.toList(),
                    selected = state.settings.language,
                    onSelect = { store.setLanguage(it) },
                    title = { it.endonym },
                    mark = { it.code.uppercase() },
                    label = strings.languageSection
                )
                Spacer(Modifier.height(10.dp))
                DropdownField(
                    options = Currency.supported,
                    selected = state.settings.currency,
                    onSelect = { store.setCurrency(it) },
                    title = { strings.currencyName(it) },
                    rowTitle = { strings.currencyRow(it) },
                    mark = { it.symbol.trim() },
                    label = strings.currencySection
                )
            }
            // The one caveat that cannot be discovered by trying it: the numbers
            // do not move when the symbol does.
            Text(
                strings.currencyNote,
                style = NocturneType.inter(12.0),
                color = Nocturne.neutral500,
                modifier = Modifier.padding(top = 8.dp)
            )

            Spacer(Modifier.height(20.dp))
            // A row, not a section. The subtitle carries the backup state,
            // because the standing reminder has to survive being folded away.
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .card()
                    .clickable(onClick = onOpenBackup)
                    .padding(13.dp)
            ) {
                Glyph(
                    if (state.settings.hasBackup) Icon.confirm else Icon.edit,
                    size = 20.dp,
                    tint = if (state.settings.hasBackup) Nocturne.accent else Nocturne.accent400
                )
                Spacer(Modifier.width(11.dp))
                Column(modifier = Modifier.weight(1f)) {
                    Text(strings.moveToAnotherPhone, style = NocturneType.rowPrimary, color = Nocturne.text)
                    Text(
                        state.settings.lastExportAt?.let { strings.backedUpOn(strings.longDate(it)) }
                            ?: strings.notBackedUpYet,
                        style = NocturneType.meta,
                        color = Nocturne.neutral500,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                }
                Glyph(Icon.openRow, size = 12.dp, tint = Nocturne.neutral600)
            }

            // Debug builds only. One tap, no confirmation, and every product,
            // price and bill is gone — right for resetting to first-run during
            // development, wrong on a counter holding the only copy of a shop.
            if (BuildConfig.DEBUG) {
                Spacer(Modifier.height(20.dp))
                Kicker(strings.startAgain, modifier = Modifier.padding(bottom = 8.dp))
                SecondaryButton(
                    strings.startOver,
                    onClick = onStartOver,
                    fullWidth = true,
                    height = 42.dp,
                    fontSize = 13.5
                )
            }
        }
    }
}

@Composable
private fun Stat(label: String, value: Int, modifier: Modifier = Modifier) {
    Column(modifier = modifier) {
        Text(label, style = NocturneType.inter(11.0), color = Nocturne.neutral500)
        Text("$value", style = NocturneType.inter(17.0), color = Nocturne.text)
    }
}
